### ISSUE TITLE
Provide an accurate way to determine if the returned response is complete or partial

### ISSUE THREAD

**Issue by Pigeo** [PostgREST/postgrest#2776]
### Environment

* PostgreSQL version: probably any
* PostgREST version: probably any
* Operating system: any

### Description of issue

The [db-max-rows](https://postgrest.org/en/stable/configuration.html#db-max-rows) is an important setting that limits the number of rows PostgREST will fetch from a view, table, or stored procedure (against accidental or malicious requests).

On my web application, I need to know wether the content returned by PostgREST is partial (`db-max-rows` has effectively limited the number of results) or complete (i.e. number of returned results < `db-max-rows`).

Since PostgREST is returning HTTP status code 206 (Partial Content) + the `Content-Range` (when passing HTTP header `Prefer: count=estimated`), I initially thought I could rely on one of these 2 hints to determine if the returned content is partial or complete, but that doesn't work, for one simple reason:

in case of [resource embedding](https://postgrest.org/en/stable/api.html#resource-embedding), the `db-max-rows` setting also applies to embedded tables (which I understand is a very good thing, for safety reasons), but the HTTP 206 doesn't apply to embedded resources (when only the embedded resource is effectively limited to `db-max-rows`, I always get an HTTP 200, even though I correctly passed the `Prefer: count=estimated` HTTP header), and the `Content-Range` is of no use here (because it doesn't give a clue about the embedded resource's rows)

### Expected behavior / suggestions of improvement

1. as long as the `db-max-rows` has effectively limited the response in *any* way (including on embedded resources), the HTTP 206 status code should always be returned (at least when `Prefer: count=estimated` HTTP header is present)
2. if you're reluctant to change the actual behaviour of the HTTP status code (for breaking-change reason), you could alternatively add another new HTTP header `Response-Is-Complete: false` that would pursue this purpose.
3. going further, it would also be nice to have size of results not only for the main table but also for each embedded resources => here I'm thinking about a small hierarchical json object, having the same hierarchical structure as the `select=` parameter of the request (with all its embedded resources), that would give us the retrieved rows / total numbers for each embedded resource (at each node of this hierarchical object).
4. another very quick and simple way to answer my initial need would simply consist of returning the `db-max-rows` value as an HTTP header of PostgREST's response, so that I can check for myself if any of the embedded ressource arrays has reached this size limit.

**Comment by Pigeo** [comment]
About suggestion #4 above, I tried to inject the `db-max-rows` value in a custom HTTP header of PostgREST's response, through [db-pre-request](https://postgrest.org/en/stable/api.html#setting-headers-via-pre-request), with the following code:
```
create or replace function custom_headers() returns void as $$
declare
  s TEXT;
begin
  SHOW pgrst.db_max_rows into s;
  perform set_config('response.headers',
      '[{"X-DB-Max-Rows": "' || s || '"}]', true);
end; $$ language plpgsql;
```
but it's not working since `pgrst.db_max_rows` is unknown... 

Any suggestion on that? 
(I tried to search for `get_config` in PostgREST's doc, by analogy with `set_config`, but found no results... Also both the [Configuration](https://postgrest.org/en/stable/configuration.html) and [Stored Procedures](https://postgrest.org/en/stable/api.html#stored-procedures) sections of the doc don't explain how to access PostgREST's configuration settings (coming from a config file) from a stored procedure.

**Comment by steve-chavez** [comment]
Sorry for the late reply here.

> in case of [resource embedding](https://postgrest.org/en/stable/api.html#resource-embedding), the db-max-rows setting also applies to embedded tables (which I understand is a very good thing, for safety reasons), but the HTTP 206 doesn't apply to embedded resources (when only the embedded resource is effectively limited to db-max-rows, I always get an HTTP 200, even though I correctly passed the Prefer: count=estimated HTTP header), and the Content-Range is of no use here (because it doesn't give a clue about the embedded resource's rows)

To aid in my understanding.

- Having a `db-max-rows=1000`.  
- Say you have a table `clients` with 500 rows.
- And you have a table `projects` with 200K rows.

You request `GET /clients?select=*,projects(*)`. So `clients` won't be limited by `db-max-rows` because it's < 1000. But `projects` will be limited. Yet you'll receive a:

```
HTTP/1.1 200 OK
```

And that is the problem because you don't know that `projects` was limited right?



> as long as the db-max-rows has effectively limited the response in any way (including on embedded resources), the HTTP 206 status code should always be returned 

So you're suggesting to always return 206 when `db-max-rows` is set? Or that we would need to count the embedded resources to return the 206 conditionally? Hm, if it's the latter that could be expensive. The former seems doable and simple to implement.

> but it's not working since pgrst.db_max_rows is unknown...
> Any suggestion on that?

pgrst.db_max_rows is not [set per-transaction](https://postgrest.org/en/stable/references/transactions.html#transaction-scoped-settings). It only shows when you use the [in-db config](https://postgrest.org/en/stable/references/configuration.html#in-database-configuration), which currently requires SUPERUSER. Can you use that?

**Comment by steve-chavez** [comment]
> Another alternative track, which can be studied in parallel:
If I was able to retrieve the value of db-max-rows from a stored procedure (or through some API endpoint provided by PostgREST), I would be able to implement the trick described above on the client side, and you wouldn't have to implement it in PostgREST's server side.

> NB: I'm not sure if I can use the in-db config: currently, I have 2 instances of PostgREST running in parallel on different SCHEMAS of the same database, with different db-max-rows values (how do you handle that with in-db config??)

@Pigeo I think with the new [pre-config](https://github.com/PostgREST/postgrest/pull/2803) you could do the above now. Different max rows and schemas can be set for 2 different instances by having 2 different authenticator roles, this can be done on the same SQL function. 

That being said, I believe we should solve this in core.

> (Personally, I would be willing to bear a small additional cost to have this behaviour, for instance by adding a new parameter in PostgREST's config to enable/disable this new functionality)

> You should internally use db-max-rows+1 when generating the SQL queries, i.e. having a db-max-rows=1000 in my PostgREST's config, you would put LIMIT 1001 in the SQL queries (instead of LIMIT 1000), and then before returning the JSON response, you would check how much rows are present in each embedded resource: if there are db-max-rows+1 rows, it means it has effectively been limited (because we asked only db-max-rows, not db-max-rows+1...), so you can return 206 (and remove the last row, in order to return db-max-rows as requested).

> What do you think about this?

Hm, that seems it could work. Kinda reminds me of the [estimated count](https://postgrest.org/en/stable/references/api/tables_views.html#estimated-count), it also does a LIMIT max-rows + 1 query underlyingly. Look:

https://github.com/PostgREST/postgrest/blob/14be3fb671b4207a1d273f9ef514fe922d40b3a1/src/PostgREST/Query.hs#L77-L81

I need to give it more thought. Maybe we can integrate that for the embedded resources.

**Comment by Pigeo** [comment]
Thanks for pointing me to the new [pre-config](https://github.com/PostgREST/postgrest/pull/2803) feature: that should solve my problem!

However I believe like you that it would be more elegant to solve this problem in core (but maybe it's not of highest priority if one can build a work-around on client's side)

**Comment by steve-chavez** [comment]
Note: the same problem was mentioned before on https://github.com/PostgREST/postgrest/issues/2156#issuecomment-1033481329

> The problem with not returning an error if db-max-rows is hit: The user doesn't know about it.

Maybe https://github.com/PostgREST/postgrest/issues/2887 could be the way forward.
